feat: add connector for QuestDB timeseries database

marregui opened the pull request:
This PR adds an Engine Spec for [QuestDB](https://github.com/questdb/questdb) a timeseries database.

The original Spec resides within module [questdb-connect](https://github.com/questdb/questdb-connect) and has been tested against Apache Superset 2.0 by the following means:

- clone this repo
- configure a secret key
- create file `docker/requirements-local.txt` with content
   ```shell
   questdb-connect>=1.0.8
   ```
- docker-compose up
- then I start a QuestDB instance [https://questdb.io/docs/](https://questdb.io/docs/)
- and I can reach it and play with it through URL `questdb://admin:quest@host.docker.internal:8812/main` defined as an OTHER connection

With this PR we would like to migrate the Engine Spec from the `questdb-connect` repo to `superset's`, as well as to enable a drop down from superset's UI so that users can select QUESTDB rightaway rather than configuring the connection through OTHER.

marregui committed:
feat: add connector for QuestDB timeseries database

villebro commented on superset/db_engine_specs/questdb.py:
```diff
@@ -0,0 +1,311 @@
+            name = builtin_time_grains[duration]
+            ret_list.append(TimeGrain(name, _(name), func, duration))
+    _engine_time_grains = tuple(ret_list)
+    _default_column_type_mappings = (
```
`_default_column_type_mappings` should not be overridden - please define `column_type_mappings` instead

marregui replied:
thank you, resolved in 9a3c588dc495ad0435666f41b455d54ee57b40fc

villebro commented on superset/db_engine_specs/questdb.py:
```diff
@@ -0,0 +1,311 @@
+        (re.compile("^BOOLEAN", re.IGNORECASE), types.Boolean, GenericDataType.BOOLEAN),
+        (re.compile("^BYTE", re.IGNORECASE), types.Byte, GenericDataType.BOOLEAN),
+        (re.compile("^SHORT", re.IGNORECASE), types.Short, GenericDataType.NUMERIC),
+        (re.compile("^INT", re.IGNORECASE), types.Int, GenericDataType.NUMERIC),
```
I believe most of these are redundant, as they're covered by the defaults in `BaseEngineSpec`. So you may want to consider only implementing whichever types are only specific to QuestDB.

marregui replied:
I am exposing all QuestDB types defined in https://github.com/questdb/questdb-connect/blob/61b9693accf390a03de73ca62e657037dd69d067/src/questdb_connect/types.py#L203

I have renamed the misleading `types.Boolean` for `questdb_types.Boolean`, etc, to make it more obvious, also in 9a3c588dc495ad0435666f41b455d54ee57b40fc

villebro commented on superset/db_engine_specs/questdb.py:
```diff
@@ -0,0 +1,311 @@
+            clause = clause.replace(":", "\\:")
+        return text(remove_public_schema(clause))
+
+    @classmethod
+    def get_time_grain_expressions(cls) -> Dict[Optional[str], str]:
+        """Return a dict of all supported time grains including any
+        potential added grains but excluding any potentially disabled
+        grains in the config file.
+        :return: All time grain expressions supported by the engine
+        """
+        return cls._time_grain_expressions
+
```
I don't think you need to override this - it should already be handled by `BaseEngineSpec`. Or is there some reason why that method doesn't work here?

marregui replied:
you are correct, tahnk you for spotting all these, addressed in the same commit 9a3c588dc495ad0435666f41b455d54ee57b40fc

villebro commented on superset/db_engine_specs/questdb.py:
```diff
@@ -0,0 +1,311 @@
+        """
+        return type_code.upper() if type_code and isinstance(type_code, str) else str(type_code)
+
+    @classmethod
+    def get_time_grains(cls) -> Tuple[TimeGrain, ...]:
+        """Generate a tuple of supported time grains.
+        :return: All time grains supported by the engine
+        """
+        return cls._engine_time_grains
```
same here

marregui replied:
and here 9a3c588dc495ad0435666f41b455d54ee57b40fc

villebro commented on superset/db_engine_specs/questdb.py:
```diff
@@ -0,0 +1,311 @@
+        """
+        return cls._engine_time_grains
+
+    @classmethod
+    def get_column_types(
+            cls,
+            column_type: Optional[str],
+    ) -> Optional[Tuple[TypeEngine, GenericDataType]]:
+        """Return a sqlalchemy native column type and generic data type that
+        corresponds to the column type defined in the data source (return None
+        to use default type inferred by SQLAlchemy). Override `column_type_mappings`
+        for specific needs (see MSSQL for example of NCHAR/NVARCHAR handling).
+        :param column_type: Column type returned by inspector
+        :return: SQLAlchemy and generic Superset column types
+        """
+        if not column_type:
+            return None
+        for regex, sqla_type, generic_type in cls._default_column_type_mappings:
+            matching_name = regex.search(column_type)
+            if matching_name:
+                return (
+                    types.resolve_type_from_name(sqla_type.__visit_name__).impl,
+                    generic_type
+                )
+        return None
```
and here

marregui replied:
correct! 9a3c588dc495ad0435666f41b455d54ee57b40fc

villebro reviewed (commented):
Thanks for adding support for QuestDB! Please check if you can remove some of those methods or properties that are already covered by `BaseEngineSpec`, but other than that this looks great 👍

marregui commented on docker/pythonpath_dev/superset_config.py:
```diff
@@ -30,6 +30,8 @@
 logger = logging.getLogger()
 
+SUPERSET_SECRET_KEY="laRamonaEsLaMasGordaDeLasMozasDeMiPuebloRamonaTeQuiero"
+SECRET_KEY=SUPERSET_SECRET_KEY
```
oops!! indeed, apologies amended in 388020b2933b4cebd70e85194e22ef5b989bec20

john-bodley commented on RELEASING/release-notes-2-0/changelog.md:
```diff
@@ -128,6 +128,7 @@ under the License.
 - [#19408](https://github.com/apache/superset/pull/19408) feat(dashboard): Implement empty states for empty tabs (@kgabryje)
 - [#19446](https://github.com/apache/superset/pull/19446) feat(explore): Move chart actions into dropdown (@kgabryje)
 - [#19394](https://github.com/apache/superset/pull/19394) feat(explore): UI changes in dataset panel on Explore page (@kgabryje)
+- [#24172](https://github.com/apache/superset/pull/24172) feat: add connector for QuestDB timeseries database (@marregui)
```
@marregui you're updating the changelog of a release that has already been shipped. The only place where we record future changes is in [UPDATING.md](https://github.com/apache/superset/blob/master/UPDATING.md), however this is used for breaking changes, which isn't the case for your PR.

john-bodley replied:
@marregui please add the dependency to `requirements/testing.in` [here](https://github.com/apache/superset/blob/master/requirements/testing.in#L19) and then you'll have to re-compile the frozen dependencies to ensure that CI is deterministic.

john-bodley commented:
Thanks @marregui for adding a new connector. Would you please mind adding [unit](https://github.com/apache/superset/tree/master/tests/unit_tests/db_engine_specs)  and [integration](https://github.com/apache/superset/tree/master/tests/integration_tests/db_engine_specs) tests for the connector?

marregui commented:
Hola,

I have added unit tests and also removed unnecesary overrides in the engine spec for QuestDB.

The tests pass locally and I have finally realised the existence of pre-commit hooks, so the code pylints well.
For some reason CI is not triggering.

I have added a dependency to https://pypi.org/project/questdb-connect/0.0.97/ which is the version of questdb-connect that does not contain the implementation of the QuestDB engine spec. There is a version 1.0.7 that simply adds a copy of the engine spec. I tested superset master with questdb-connect 0.0.97, it works well afaict. 

Regards

marregui commented:
hi, could I please have a review?

eschutho commented:
Hi @marregui sorry that this PR fell through the cracks. Happy to pick up on this and help you get it merged. Can you rebase your PR?

rusackas commented:
This has sat untouched in draft mode for quite a while, so we'll go ahead and close it. We welcome you to reopen it if you want to contribute the change. Thank you for your contribution nonetheless!
